packages/walt-compiler/src/tokenizer/comments/index.js

```diff
@@ -2,17 +2,49 @@
 import token from "../token";
 import Syntax from "../../Syntax";
 
-const everything = () => everything;
+import {
+  SLASH,
+  SINGLE_LINE,
+  MULTI_LINE_START,
+  MULTI_LINE_END,
+  COMMENT_IDENTIFIERS,
+  everything,
+} from "./constants";
 
-const slash = char => {
-  if (char === "/") {
-    return everything;
+let isMultiline: boolean = false;
+let previous: string;
```

Comment:
So this is the last issue I have with the changes here, and that is these two dependencies on module scoped variables.

The tokenizer should be as pure as possible, but these two variables make that impossible/difficult. Honestly, this is a theoretical problem only, but I'd still like to be as cautious as possible.

Instead of defining them in the module, have the `maybeComment` function return a closure to avoid the issue.

Kind of like this

```js
const maybeComment = (current: string) => {
   let isMultiline: boolean = false;
   let previous: string;

   const isComment = (current: string) => {
      ...
   }

  // your maybe comment logic here
}
```

This way we make sure that there are no surprise side-effects of running this code (like one parsed comment affecting the next). I think this should just work based on your change already.

Reply:
I'll give it a go.

Reply:
I did try to do this, but i need the `isMultiline` in between maybeComment calls and `previous` as well.I'm not seeing the way that it can be done, but i've done something else that i believe to be even better (since isComment isn't created on every maybeComment call).

```javascript
// @flow
import token from "../token";
import Syntax from "../../Syntax";

import {
  SLASH,
  SINGLE_LINE,
  MULTI_LINE_START,
  MULTI_LINE_END,
  COMMENT_IDENTIFIERS,
  everything,
} from "./constants";

const parser = () => {
  let isMultiline: boolean = false;
  let previous: string;

  const maybeComment = (current: string) => {
    const buffer: string = previous;
    previous = current;

    const isComment = (char: string) => {
      switch (`${previous}${char}`) {
        case MULTI_LINE_END: {
          isMultiline = false;
          return isComment;
        }
        case MULTI_LINE_START: {
          isMultiline = true;
          return isComment;
        }
        case SINGLE_LINE: {
          return everything;
        }
        default: {
          if (isMultiline) {
            previous = char;
            return isComment;
          }
        }
      }
    };

    if (current === SLASH || isMultiline ||
      COMMENT_IDENTIFIERS.indexOf(`${buffer}${current}`) > -1) {
      return isComment;
    }

    return null;
  };

  return token(maybeComment, Syntax.Comment);
};

const commentParser = parser();
export default commentParser;
```

Reply:
@ballercat Let me know what you think.

Reply:
Ah yeah, I see what you mean. That is about right. Except you need to tweak it a bit by not doing `const commentParser = parser();`, because then we just end up with a single closure again. We want the closure to be re-created every time `parser` function is called to avoid unintentional side-effects.

Inside the `parser` function adjust it to call `maybeComment` the final thing should look kind of like so:

```js
const parser = (char) {
  // all the other declarations here as before
  ...
  return maybeComment(char);
};
export default token(parser, Syntax.Comment);
```

Reply:
@ballercat Tried to do what you said, but the `parser` function is called for every single token making it impossible to retain the previous value. So i think having a single closure isn't that big of a deal here. Unless there's something else i don't see, would you accept the PR as is? (I've committed the changes from above)

Reply:
> function is called for every single token making it impossible to retain the previous value

The Token parser is intended to be Finite State Machine, where each token parser is called once when we begin parsing. Each parser returns a new function which can match the next valid state, or null if no match is possible. All of the parsers are in an array which we `.filter` and remove ones that cannot produce a match, once a single parser is left we know we have a match.

I think the difficulty with multi-line comments here is that the logic for the parser resets on every new line, and previously there have been no tokens which can persist across multiple lines. I'll need to take a closer look at it sometime after work.

Reply:
@whitecrownclown 

I identified a problem in the tokenizer which is the problem. As I thought it's the fact that we handle end of line characters differently from all the rest(we ignore/skip over them).

Simply remove the `eol` check here:
https://github.com/ballercat/walt/blob/f042bad67b9b68e799bfc9f031022741a8bee83d/packages/walt-compiler/src/tokenizer/index.js#L63

And the closure approach will work as I described(I quickly tested it). I think you will need to tweak single line comment logic to break on newline characters, but other than that it should work. Identifier token parser might need to be tweaked as well to break on newline.

Reply:
@ballercat I see, thanks for looking into this, i'll go ahead on this as soon as i can and get back to you.